added a fake bms with set_charging and set_relative_remaining_capacity services

souravran opened the pull request.

souravran committed:
added a fake bms with set_charging and set_relative_remaining_capacity services

fmessmer commented on cob_bms_driver/src/fake_bms.py:
```diff
@@ -0,0 +1,39 @@
+
+import rospy
+from cob_msgs.msg import PowerState
+from cob_srvs.srv import SetFloat, SetFloatResponse
```
please add 
```
<exec_depend>cob_msgs</exec_depend>
<exec_depend>cob_srvs</exec_depend>
```
to `cob_bms_driver/package.xml` (respect alphabetical order :wink:)

fmessmer commented on cob_bms_driver/src/fake_bms.py:
```diff
@@ -0,0 +1,39 @@
+    def __init__(self):
+        self.srv_charging = rospy.Service('~set_charging', SetBool, self.charging_cb)
+        self.srv_relative_remaining_capacity = rospy.Service('~set_relative_remaining_capacity', SetFloat, self.relative_remaining_capacity_cb)
+        self.pub_power_state = rospy.Publisher('/power_state', PowerState, queue_size = 1)
```
I thought, we discussed to publish 6 separate topics `voltage`, `current`, `remaining_capacity`, `full_charge_capacity`, `temperature`, `battery_charging` - as the real bms driver does - instead of publishing the aggregated `/power_state`
Because then we can use the same `power_state_aggregator.py` node to aggregate the `/power_state` topic in simulation and in real hw!

fmessmer commented:
also, please add install tags

souravran committed:
fake_bms publishing all power_state entities.
added exception handling in power_state_aggregator.
added package dependency and install tags.

fmessmer commented on cob_bms_driver/src/fake_bms.py:
```diff
@@ -0,0 +1,51 @@
+        self.temperature        = 0.0
+        self.charging_state     = False
+
+        rospy.Timer(rospy.Duration(0.1), self.timer_cb)
```
can we use same poll/publish frequency as in the real driver? (see parameter https://github.com/ipa320/cob_driver/blob/indigo_dev/cob_bms_driver/src/cob_bms_driver_node.cpp#L180)

fmessmer commented on cob_bms_driver/src/power_state_aggregator.py:
```diff
@@ -62,11 +62,15 @@ def calculate_power_consumption(self):
-        else:
-            rospy.logwarn("cannot calculate relative remaining capacity. full_charge_capacity=%d, remaining_capacity=%d" % (self.full_charge_capacity, self.remaining_capacity))
-            return 0.0
+        print self.remaining_capacity, "----", self.full_charge_capacity
```
remove or `rospy.logdebug`

fmessmer commented on cob_bms_driver/src/power_state_aggregator.py:
```diff
@@ -62,11 +62,15 @@ def calculate_power_consumption(self):
+        try:
+            ret_value = round(100.0*(self.remaining_capacity/self.full_charge_capacity), 3)
+        except ZeroDivisionError as e:
+            rospy.logwarn("invalid relative remaining capacity data :%s" % (e))
```
`rospy.logerror("ZeroDivisionError: full_charge_capacity is 0.0: %s" % (e))`

fmessmer reviewed (changes requested):
minor requests

fmessmer commented:
fixes https://github.com/ipa320/cob_docking/issues/17

souravran committed:
made changes as per the review.
power state elements being published at 20 Hz.
removed junk rospy log and changed division_by_zero error message.

fmessmer commented on cob_bms_driver/src/fake_bms.py:
```diff
@@ -0,0 +1,51 @@
+        self.temperature        = 0.0
+        self.charging_state     = False
+
+        rospy.Timer(rospy.Duration(0.2), self.timer_cb)
```
as discussed, please use the `poll_frequency_hz` parameter here

souravran committed:
poll frequency has been set from the parameter list

fmessmer commented on cob_bms_driver/src/fake_bms.py:
```diff
@@ -10,12 +10,13 @@ class FakeBMS(object):
-        self.pub_charge_capacity    = rospy.Publisher('/bms/full_charge_capacity', Float64, queue_size = 1)
-        self.pub_temparature        = rospy.Publisher('/bms/temperature', Float64, queue_size = 1)
-        self.pub_charging_state     = rospy.Publisher('/bms/battery_charging', Bool, queue_size = 1)        
+        self.poll_frequency          = rospy.get_param('poll_frequency_hz')
```
Use the `default_param = rospy.get_param('default_param', 'default_value')` notation:

`self.poll_frequency         = rospy.get_param('poll_frequency_hz', 20.0)`

check/beautify indentation :wink:

fmessmer commented on cob_bms_driver/src/fake_bms.py:
```diff
@@ -24,8 +25,15 @@ def __init__(self):
 
-        rospy.Timer(rospy.Duration(0.2), self.timer_cb)
-        rospy.loginfo('FakeBMS is running')
+        try:
```
`try-except` should not be necessary in case you use the `default_value` notation

fmessmer reviewed (changes requested):
almost

souravran committed:
uses the default parameter value

souravran committed:
harmonize namespaces of fake_bms

souravran committed:
publish diagnostics in fake_bms

souravran committed:
consistent naming

souravran committed:
fake current

souravran committed:
finalize exception handling

fmessmer reviewed (approved):
I finalized it together with @souravran
